File: packages/cli/commands/duplicate.ts
import { deconflictHostPorts, duplicateInstance } from "@compositz/core";
import { dim, green, red, yellow } from "@std/fmt/colors";
import { storeDir } from "../lib.ts";

/** Derive a fresh instance from an existing one (copies the bundle, not the data). */
export async function duplicateCmd(args: string[]): Promise<number> {
  if (!args[0]) {
    console.error(red("usage: compositz duplicate <instanceId>"));
    return 1;
  }
  const instance = await duplicateInstance(storeDir(), args[0]);
  // A duplicate shares the source recipe's ports, so it always collides — reassign + report.
  for (const b of await deconflictHostPorts(storeDir(), instance)) {
    console.log(yellow(`  note: ${b.name} port ${b.from} in use → reassigned to ${b.to}`));
  }
  console.log(green("OK — duplicated") + dim(` ${args[0]} → ${instance.instanceId}`));
  console.log(dim(`  run: compositz up ${instance.instanceId}`));
  return 0;
}

File: crates/cli/src/commands/duplicate.rs
//! `duplicate` — derive a fresh instance from an existing one (copies the bundle,
//! not the data), then deconflict the copy's host ports.

use anyhow::Result;
use compositz_core::{Instance, deconflict_host_ports, duplicate_instance, is_valid_instance_id};

use crate::cli::store_dir;
use crate::style::{dim, green, yellow};

/// Duplicate an instance's definition under a new id. The source id is validated at
/// the boundary (★ F5) before it flows into the copy (which loads + walks the
/// source dir).
pub async fn run(instance_id: String) -> Result<i32> {
    if !is_valid_instance_id(&instance_id) {
        anyhow::bail!("invalid instance id: \"{instance_id}\"");
    }
    let store = store_dir()?;

    let source = instance_id.clone();
    let store_for_copy = store.clone();
    let instance: Instance = tokio::task::spawn_blocking(move || -> Result<Instance> {
        Ok(duplicate_instance(&store_for_copy, &source)?)
    })
    .await??;

    // A duplicate shares the source recipe's ports, so it always collides — reassign
    // + report.
    for bump in deconflict_host_ports(&store, &instance)? {
        println!(
            "{}",
            yellow(&format!(
                "  note: {} port {} in use → reassigned to {}",
                bump.name, bump.from, bump.to
            ))
        );
    }

    println!(
        "{}{}",
        green("OK — duplicated"),
        dim(&format!(" {} → {}", instance_id, instance.instance_id))
    );
    println!(
        "{}",
        dim(&format!("  run: compositz up {}", instance.instance_id))
    );
    Ok(0)
}
